docs/language/learn-ql/javascript/type-tracking.rst

```diff
@@ -0,0 +1,523 @@
+- To ensure the initial value is a source node, a call to ``getALocalSource()`` is usually required.
+
+.. code-block:: ql
+
+  SourceNode firebaseSnapshotCallback(string refName, TypeBackTracker t) {
+    t.start() and
+    result = firebaseRef(refName).getAMethodCall("once").getArgument(1).getALocalSource()
+    or
+    exists(TypeBackTracker t2 |
+      result = firebaseSnapshotCallback(refName, t2).backtrack(t2, t)
+    )
+  }
+
+  FunctionNode firebaseSnapshotCallback(string refName) {
+    result = firebaseSnapshotCallback(refName, TypeBackTracker::end())
+  }
+
+Now, ``firebaseSnapshotCallback("forecast")`` finds the function being passed to ``fetchForecast``.
+Based on that we can track the ``snapshot`` value and find the ``val()`` call itself:
+
+.. code-block:: ql
+
+  SourceNode firebaseSnapshot(string refName, TypeTracker t) {
+    t.start() and
+    result = firebaseSnapshotCallback(refName).getParameter(0)
+    or
+    exists(TypeTracker t2 |
+      result = firebaseSnapshot(refName, t2).track(t2, t)
+    )
+  }
+
+  SourceNode firebaseSnapshot(string refName) {
+    result = firebaseSnapshot(refName, TypeTracker::end())
+  }
+
+  MethodCallNode firebaseDatabaseRead(string refName) {
+    result = firebaseSnapshot(refName).getAMethodCall("val")
+  }
+
+With this addition, ``firebaseDatabaseRead("forecast")`` finds the call to ``snapshot.val()`` which contains the value of the forecast.
+
+`Here <https://lgtm.com/query/8761360814276109092>`__ is a run of an example query using the model to find `val` calls.
+
+Summary
+-------
+
+This covers the use of the type tracking library. To recap, use this template to define forward type tracking predicates:
+
+.. code-block:: ql
+
+  SourceNode myType(TypeTracker t) {
+    t.start() and
+    result = /* SourceNode to track */
+    or
+    exists(TypeTracker t2 |
+      result = myType(t2).track(t2, t)
+    )
+  }
+
+  SourceNode myType() {
+    result = myType(TypeTracker::end())
+  }
+
+Use this template to define backward type tracking predicates:
+
+.. code-block:: ql
+
+  SourceNode myType(TypeBackTracker t) {
+    t.start() and
+    result = (/* argument to track */).getALocalSource()
+    or
+    exists(TypeBackTracker t2 |
+      result = myType(t2).backtrack(t2, t)
+    )
+  }
+
+  SourceNode myType() {
+    result = myType(TypeBackTracker::end())
+  }
+
+Note that these predicates all return ``SourceNode``,
+so attempts to track a non-source node, such as an identifier or string literal,
+will not work.
+
+Also note that the predicates taking a ``TypeTracker`` or ``TypeBackTracker`` can often be made ``private``,
+as they are typically only used as an intermediate result to compute the other predicate.
+
+Limitations
+-----------
+
+As mentioned, type tracking will track values in and out of function calls and properties,
+but only within some limits.
+
+Type tracking does not always track *through* functions, that is, if a value flows into a parameter
```

Comment:
Is it unrealistic that we one day add this feature to type tracking?

```suggestion
For example, type tracking does not always track *through* functions at the moment, that is, if a value flows into a parameter
```

Reply:
In general, it's not possible at all. We can add steps for the simple cases where the entire function can be summarized as a single store/load or plain step, but unlike data flow configurations, the summarization can never depend on any custom steps. All type tracking predicates must share the same set of function summaries.